pkg/controllers/managementuser/resourcequota/resource_quota_sync.go

```diff
@@ -244,39 +238,46 @@ func (c *SyncController) getExistingLimitRange(ns *corev1.Namespace) (*corev1.Li
 	return limitRanger[0], nil
 }
 
-func (c *SyncController) getNamespaceResourceQuota(ns *corev1.Namespace) (*corev1.ResourceQuotaSpec, error) {
-	limit, err := getNamespaceResourceQuotaLimit(ns)
+// deriveRequestedResourceQuota tries to obtain the new namespace's resource quota limit and its quota spec.
+// It derives it by looking up the requested quota limit. If it's not found, then it looks up the project's default
+// quota for a namespace. If it's also not found, then the method returns an error.
+// If only the requested quota limit exists, then it is returned. If only the project's default namespace limit exists,
+// then it is returned. If both exist, then the two limits are merged, with requested limits having priority.
+func (c *SyncController) deriveRequestedResourceQuota(ns *corev1.Namespace) (*v32.ResourceQuotaLimit, *corev1.ResourceQuotaSpec, error) {
+	requested, err := getNamespaceResourceQuotaLimit(ns)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	if limit == nil {
-		limit = defaultResourceLimit
+
+	defaultQuota, err := getProjectNamespaceDefaultQuota(ns, c.ProjectLister)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	return convertResourceLimitResourceQuotaSpec(limit)
-}
+	var quotaLimit *v32.ResourceQuotaLimit
 
-func getDefaultQuotaSpec() (*corev1.ResourceQuotaSpec, error) {
-	return convertResourceLimitResourceQuotaSpec(defaultResourceLimit)
-}
+	if requested != nil && defaultQuota == nil {
+		quotaLimit = requested
```

Comment:
Is this correct? Is this the behavior that happened before this change? It looks like the UI doesn't even let you set a quota on the namespace if the project has no default.

Reply:
Right, the UI lets you set the quota only for those resources that are found in the project quota. The necessity of this, by the way, is a question for the future.
But this handles both UI and kubectl. In the manifest users can specify anything in the annotation.
And, yes, this behavior is not changing.

Reply:
This actually does seem to be a change in behavior from my testing. Without this change, if you create a project with no default quota set and a namespace like

```
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    field.cattle.io/projectId: local:p-9zk9c
    field.cattle.io/resourceQuota: '{"limit":{"pods":"50","limitsCpu":"1000m"}}'
  name: n4
spec:
```

then no ResourceQuota would be created in the namespace. With this change, there now is one. I don't think we should make any further changes to the quota behavior unrelated to the original bug without further discussion.

Reply:
Sorry, I misunderstood your original question - I thought you were asking about resources not defined on the project and their quota vs that for resources that ARE defined on the project.

You're right, the behavior does change, I wasn't aware of it, my fault. I believe though that the new behavior is more intuitive, but, as you said, we need to discuss this.
@cbron

Reply:
It all comes down to the role of the annotation and whether users should use it and rely on it. If we maintain that it should not be used, then the old behavior makes more sense. But if we honor the annotation, then the new behavior should be expected.

Reply:
Discussed in standup - for this case, reverting to existing behavior. If there is a requested quota but no default on the project, then do not apply quota limits.